### PR TITLE
feat: improve help message for `new` command

### LINKED ISSUE #162: Improve help message with list of available examples
#### Reason/Context
Doing `asyncapi new --file-name=asyncapi.yml --example=default-example.yaml --no-tty` us super cool, but to get `default-example.yaml` you need to navigate to CLI installation, go into `assets/examples` and see the list of examples, or check if they will be listed in the interactive mode. You also need to kind of guess if example name is just the file name or filename with its extension.

So great feature is kind of hidden from users.

#### Description

Now I get:
```
$  asyncapi new --help

creates a new asyncapi file

USAGE
  $ asyncapi new

OPTIONS
  -e, --example=example      name of the example to use
  -h, --help                 show CLI help
  -n, --file-name=file-name  name of the file
  -p, --port=port            port in which to start Studio
  -s, --studio               open in Studio
  --no-tty                   do not use an interactive terminal
```

I should get:
```
$  asyncapi new --help

creates a new asyncapi file

USAGE
  $ asyncapi new

OPTIONS
  -e, --example=example.fileExtension      name of the example to use. Available examples are:
                                                - default-example.yml
                                                - somethingElseWithSpecificProtocol.json (MQTT)
  -h, --help                 show CLI help
  -n, --file-name=file-name  name of the file
  -p, --port=port            port in which to start Studio
  -s, --studio               open in Studio
  --no-tty                   do not use an interactive terminal
```

Bonus would be to see also example for `asyncapi new` at the bottom of help like:
```
EXAMPLES
  asyncapi new               start creation of a file in interactive mode
  asyncapi new --file-name=my-asyncapi.yml --example=default-example.yml --no-tty        create new file with specific name, using one of examples and without interactive emode

### PR DESCRIPTION
**Pull request by LakshyaSatpal**
**Description**

- Improved the help message of 'new' command with examples

**Related issue(s)**

Fixes #162

### CODE REVIEW

File: src/commands/new.ts [hunk 1]
```diff
@@ -15,14 +15,22 @@ export default class New extends Command {
   static flags = {
     help: Flags.help({ char: 'h' }),
     'file-name': Flags.string({ char: 'n', description: 'name of the file' }),
-    example: Flags.string({ char: 'e', description: 'name of the example to use' }),
+    example: Flags.string({ char: 'e', description: `name of the example to use. Available examples are: \n\t- default-example.yml \n\t- somethingElseWithSpecificProtocol.json (MQTT)` }),
```
[the remark this answers is not in the document]
Comment (LakshyaSatpal) [reply src/commands/new.ts:18]: Okay, got it, so basically I've to import that JSON file and make the description and examples from that array, right?

File: src/commands/new.ts [hunk 2]
```diff
@@ -24,6 +46,10 @@ export default class New extends Command {
   static args = []
 
   async run() {
+		const exampleDescription = await getExamplesFlagDescription();
+		this.flags.example = Flags.string({ char: 'e', description: exampleDescription});
```
Comment (LakshyaSatpal) [inline src/commands/new.ts:50]: Okay, so it's been a while now that I'm stuck on this.
The problem that I'm facing is that I fetched the example.json, mapped it, and made the description out of it. But, how do I add to my `static flags`? 
I've tried to add it inside the run() function, before parsing the `New` class, but it's not working.

Comment (derberg) [reply src/commands/new.ts:50]: I think flag building doesn't like async/await and promises 😄 

when I quickly play with your code:
- replace any async/await with normal functions and switch to `readFileSync`
- fix building description like this:
  ```
  let description = 'name of the example to use. Available examples are:';
 
  for (const element of examples) {
    description += (`\n\t - ${element.value}`);
  }

it works as expected:
<img width="959" alt="Screenshot 2022-07-19 at 14 17 01" src="https://user-images.githubusercontent.com/6995927/179748224-8b68ef39-2eb2-4e12-af2e-098c5eba247d.png">


full code:

```
function loadExampleFile() : Example[] {
  const exampleFiles = readFileSync(resolve(__dirname, '../../assets/examples/examples.json'), { encoding: 'utf8' });
  //console.log(JSON.parse(exampleFiles));
  return JSON.parse(exampleFiles);
}

function getExamplesFlagDescription() : string {
  const examples = loadExampleFile();
  let description = 'name of the example to use. Available examples are:';
 
  for (const element of examples) {
    description += (`\n\t - ${element.value}`);
  }

  return description;
}
```

```
    example: Flags.string({ char: 'e', description: getExamplesFlagDescription()}),
```

Comment (LakshyaSatpal) [reply src/commands/new.ts]: Got it, thanks a lot

**Comment by derberg** [comment]
@LakshyaSatpal looking good man! Please just add a unit test that validates if list of examples is rendered? I'm just pretty sure that in few months someone will say "why do we use sync function here, lemme refactor" 😆 

in the test, please do not "hardcode" list of examples as the tests will be affected by changes in examples list. So in test also read examples dynamically

**Comment by LakshyaSatpal** [comment]
> 

Hi!
I was going through oclif documentation for testing for commands. I'm finding it challenging to write a test to check if a file is rendered. 
I mean, I couldn't get what do I have to actually test, validate if the user entered the correct example name from available examples? or only check the working of loadExampleFile synchronous function that we wrote?
Btw, this is the first time I am writing unit tests. Please bear with me 😅

**Comment by derberg** [comment]
@LakshyaSatpal no worries. Happy to help you if you need.

- you need to provide test for new functionality. New functionality is: much better help message with list of examples. We need to have a test, that will evaluate if the list of examples from `/assets/examples` is included in the help message that you just created.
- have a look on existing tests https://github.com/asyncapi/cli/tree/master/test/commands for commands. See how we check if certain message appears in the console

I recommend, especially if it is your first time, first create a test, where you basically hardcode the expected message in the text, so you check if the message printed in terminal match the message that you hardcode in test (with list of examples).

Once ☝🏼 works, then work on making sure that the list of examples is not hardcoded, but you read in dynamically inside the test, from examples dir

**Comment by LakshyaSatpal** [comment]
@derberg I wrote this simple test

**Comment by LakshyaSatpal** [comment]
@derberg I wrote this simple test where I checked for help message to contain the first example which I hard coded.
```
test
		.stdout()
		.command(['new', '--help'])
		.it('help includes available examples', (ctx, done) => {
			expect(ctx.stdout).to.contain('- simple.yml');
			done();
		});
```
This shows me an error of `AssertionError: expected '' to include '- simple.yml'`. I believe, it is not getting that I'm checking for help flag command.

File: test/commands/new.test.ts [hunk 3]
```diff
@@ -20,4 +20,12 @@ describe('new', () => {
       expect(ctx.stdout).to.equals('Created file specification.yaml...\n');
       done();
     });
+
+	test
+		.stdout()
+		.command(['new', '-h'])
```
Comment (Souvikns) [inline test/commands/new.test.ts:26]:
```suggestion
		.command(['new', '--help'])
```

try doing this

Comment (LakshyaSatpal) [reply test/commands/new.test.ts:26]: Yes, I tried, getting the same error

Comment (Souvikns) [reply test/commands/new.test.ts:26]: Yeah, not working only in the test environment 🤔

Comment (Souvikns) [reply test/commands/new.test.ts:26]: found this though https://github.com/oclif/oclif/issues/142#issuecomment-406817611

**Comment by derberg** [comment]
@LakshyaSatpal hey, you need a hand? I was out on holidays, but now back 💪🏼

**Comment by LakshyaSatpal** [comment]
> @LakshyaSatpal hey, you need a hand? I was out on holidays, but now back 💪🏼

Hi @derberg, still stuck now the same problem - not able to write a test with help flag 😥, discussed it with @Souvikns as well

File: test/commands/new.test.ts (hunk 3 above)
Comment (derberg) [reply test/commands/new.test.ts:26]: the reason is that actually `new --help` in test is not successful and throws:

```js
TypeError: cmd._help is not a function
```

I get the same when I try `bin/run new -h` with CLI. But `bin/run new --help` works fine when I try your branch 🤔

Comment (derberg) [reply test/commands/new.test.ts:26]: @Souvikns maybe we should just resign from tests in this PR and write an issue about what I described above and get that solved first. Thoughts?

Comment (derberg) [reply test/commands/new.test.ts:28]: ok, we need to go forward here:
@LakshyaSatpal please look at other comments, address them, and remove the test. There is no way to make it run now. The main functionality will have to stay untested 😄 

also please add `"additionalHelpFlags": ["-h"],` to oclif configuration in `package.json` (line 97) so we at least make `-h` to work too.

**Comment by derberg** [comment]
@LakshyaSatpal I hope you are still interested with this one. It helps with DX and as you see it is ok that we drop testing part

**Comment by LakshyaSatpal** [comment]
@derberg Okay, I removed the test and addressed other feedbacks. Please see if we can merge now

**Review by derberg — DISMISSED** [review]
looks great from my point of view. Tried locally and works like a charm

```
(improve-new-help-msg)$ bin/run new -h
Creates a new asyncapi file

USAGE
  $ asyncapi new [-h] [-n <value>] [-e <value>] [-s] [-p <value>] [--no-tty]

FLAGS
  -e, --example=<value>
      name of the example to use. Available examples are:
        - simple.yml
        - anyof.yml
        - application-headers.yml
        - correlation-id.yml
        - websocket-gemini.yml
        - gitter-streaming.yml
        - mercure.yml
        - not.yml
        - operation-security.yml
        - oneof.yml
        - rpc-client.yml
        - rpc-server.yml
        - slack-rtm.yml
        - tutorial.yml
        - streetlights-kafka.yml
        - streetlights-operation-security.yml
        - streetlights-mqtt.yml

  -h, --help
      Show CLI help.

  -n, --file-name=<value>
      name of the file

  -p, --port=<value>
      port in which to start Studio

  -s, --studio
      open in Studio

  --no-tty
      do not use an interactive terminal

DESCRIPTION
  Creates a new asyncapi file

EXAMPLES
  $ asyncapi new         - start creation of a file in interactive mode

  $ asyncapi new --file-name=my-asyncapi.yml --example=default-example.yml --no-tty      - create a new file with a specific name, using one of the examples and without interactive mode
```

@Souvikns wanna have a look?

**Comment by derberg** [comment]
@Souvikns ping 😉

**Review by Souvikns — APPROVED** [review]
LGTM! just need to fix lint

File: test/commands/new.test.ts (hunk 3 above)
Comment (Souvikns) [reply test/commands/new.test.ts]: Yeah, let's not add tests in this PR, but open an issue about how we cannot test the help messages. And once oclif fixes it we can add help tests for all other commands.

**Comment by derberg** [comment]
@LakshyaSatpal hay can you fix linter issues please

**Comment by derberg** [comment]
I have no idea why tests are failing 😭 
it is not random, it is always `[ERR_INVALID_ARG_TYPE]: The "warning" argument must be of type string or an instance of Error. Received an instance of TypeError` but why.......#$%

@LakshyaSatpal maybe you could switch to `readFile` that is already used in the same file, instead of importing `readFileSync`? or was sync selected because `description: await getExamplesFlagDescription()` was not possible?

**Comment by derberg** [comment]
@LakshyaSatpal not sure why but you are having a lot of linter issues. Do you want to continue with this PR?

**Comment by derberg** [comment]
@LakshyaSatpal can you specify if you plan to continue with this one, or should that be released to other contributors?

**Comment by derberg** [comment]
no reaction so closing, hope someone will pick it up from https://github.com/asyncapi/cli/issues/162

**Comment by derberg** [comment]
done with https://github.com/asyncapi/cli/pull/758

**Pull request closed without merging**